File: 2271-rearrange-array-elements-by-sign/2271-rearrange-array-elements-by-sign.cpp
class Solution {
public:
    vector<int> rearrangeArray(vector<int>& nums) {
        // Brute force will be using extra space O(n) and time O(2n)

        int n = nums.size();
        vector<int>pos;
        vector<int>neg;

        for(int i = 0; i < n; i++){
            if(nums[i] > 0)pos.push_back(nums[i]);
            else neg.push_back(nums[i]);
        }

        int k = 0;
        for(int i = 0; i < n/2; i++){
            nums[k] = pos[i];
            nums[k+1] = neg[i];
            k+=2;
        }

        return nums;

        // optimized will be in the constant space
    }
};

File: 2271-rearrange-array-elements-by-sign/2271-rearrange-array-elements-by-sign.java
class Solution {
    public int[] rearrangeArray(int[] nums) {
        // Naive Approach -> O(n) and Space -> O(2*n)
        // We can reduce the Space as well
        int n = nums.length;
        int[] pos = new int[n/2];
        int[] neg = new int[n/2];
        int i = 0,j = 0;

        for(int k = 0; k < n; k++){
            if(nums[k] > 0){
                pos[i++] = nums[k];
            }else{
                neg[j++] = nums[k];
            }
        } 

        int[] res = new int[n];
        int h = 0;

        for(i = 0; i < n/2; i++){
            res[h++] = pos[i];
            res[h++] = neg[i]; 
        }

        return res;
    }
}